src/cressi_goa.c

```diff
@@ -474,21 +597,70 @@ cressi_goa_device_foreach (dc_device_t *abstract, dc_dive_callback_t callback, v
-			goto error_free_dive;
+		if (conf->version == 0) {
+			// Verify the header in the logbook and dive data are identical.
+			// After the 2 byte dive number, the logbook header has 5 bytes
+			// extra, which are not present in the dive header.
+			if (dive_size < SZ_HEADER - 5 ||
+				memcmp (dive_data + 0, logbook_data + offset + 0, 2) != 0 ||
+				memcmp (dive_data + 2, logbook_data + offset + 7, SZ_HEADER - 7) != 0) {
+				ERROR (abstract->context, "Unexpected dive header.");
+				status = DC_STATUS_DATAFORMAT;
+				goto error_free_dive;
+			}
+		} else {
+			/* The header format is as follows:
+			 *
+			 *  ID  | SAMPLES | START | RATE | TARA | SESSION | DIPS | unused | ...
+			 *  u16 |   u16   |  DATE |  u8  |  u8  |   u16   |  u8  |   u8   | ...
+			 *
+			 *  ... MAXDEPTH | MINTEMP | SURFTIME | DIVETIME | BESTDIP | CRC
+			 *  ...    u16   |   u16   |    u16   |    u16   |   u16   | u16
+			 *
+			 *  START = A Date in the "default" byte-wise form u16u8u8u8u8 for YMDhm
+			 *  RATE = SampleRate - 1=0.5sec 2=1sec 3=2sec
+			 *  TARA = Taravana Factor
+			 *  SESSION = Session Time
+			 *  DIPS = Number of Dips/Dives in this session
+			 */
+			unsigned short log_entry = array_uint16_le(logbook_data + offset);
+			unsigned short log_entry_ = array_uint16_le(dive_data);
+			if (log_entry_ != log_entry) {
+				ERROR (abstract->context, "Unexpected log entry %d != %d.", log_entry_, log_entry);
+				status = DC_STATUS_DATAFORMAT;
+				goto error_free_dive;
+			}
+			const unsigned char *start_date = logbook_data + offset + 3;
+			if (memcmp(dive_data + 4, start_date, 6) != 0) {
+				unsigned long long is = array_uint64_be(dive_data + 4);
+				unsigned long long should = array_uint64_be(start_date);
+				is >>= 16;
+				should >>= 16;
+				ERROR (abstract->context, "Unexpected start date 0x%0llx != 0x%0llx.", is, should);
+				status = DC_STATUS_DATAFORMAT;
+				goto error_free_dive;
+			}
+			unsigned short num_dips = dive_data[14];
+			DEBUG (abstract->context, "Received %zu bytes of data for dip %d with %d samples.",
+					dc_buffer_get_size (dive),
+					array_uint16_le(dc_buffer_get_data (dive)),
+					num_dips);
 		}
 
-		// Those 5 extra bytes contain the dive mode, which is required for
-		// parsing the dive data. Therefore, insert all 5 bytes again. The
-		// remaining 4 bytes appear to be some 32 bit address.
-		if (!dc_buffer_insert (dive, 2, logbook_data + offset + 2, 5)) {
+		/* Inject a header
+		 * [0] = divecomputer  -  0xdc
+		 * [1] = divecomputer  -  0xdc
+		 * [2] = version       -  0x**
+		 * [3] = length        -  0x**
+		 */
+		unsigned char header[4] = {0xdc, 0xdc, conf->version, conf->logbook_entry_len};
+		if (!dc_buffer_insert (dive, 0, header, 4)) {
+			ERROR (abstract->context, "Out of memory.");
+			status = DC_STATUS_NOMEMORY;
+			goto error_free_dive;
+		}
+		/* Inject the logbook entry, which contains the dive mode */
+		if (!dc_buffer_insert (dive, 4, logbook_data + offset, conf->logbook_entry_len)) {
```

Comment:
What do you think of this approach to keep the relevant data together?

Reply:
I need to look into these details further. I have been concentrating on the communication part first, because we have to return our test device soon. (We loaned one from a user for the BLE communication, but for your work that's now a nice coincidence too!) For the data part, not having physical access to a test device is less problematic.